src/uproot/_util.py

```diff
@@ -809,21 +809,32 @@ def objectarray1d(items):
 
 
 def regularize_steps(steps):
-    if not isinstance(steps, (list, set)):
-        raise TypeError("The specification of steps should be a list or a set.")
-
-    out = steps
-    if isinstance(steps, set):
-        out = list(steps)
     out = numpy.array(steps)
 
-    if len(out.shape) > 2 or out.dtype not in [numpy.int64, numpy.int32]:
-        raise ValueError(
-            "steps should be specified as a list of integer offsets or a list of pairs of integer starts and stops."
+    if isinstance(steps, dict) or not issubclass(out.dtype.type, numpy.integer):
+        raise TypeError(
+            "'files' argument's steps must be an iterable of integer offsets or start-stop pairs."
+        )
+
+    if len(out.shape) == 1:
+        if len(out) == 0 or not numpy.all(out[1:] >= out[:-1]):
+            raise ValueError(
+                "if 'files' argument's steps are (one-dimensional) offsets, they must be non-empty and monotonically increasing"
+            )
+
+    elif len(out.shape) == 2:
+        if not all(out[:, 1] >= out[:, 0]):
+            raise ValueError(
+                "if 'files' argument's steps are (two-dimensional) start-stop pairs, all stops must be greater than or equal to their corresponding starts"
+            )
+
+    else:
+        raise TypeError(
+            "'files' argument's steps must be an iterable of integer offsets or a list of pairs of integer starts and stops."
         )
 
     if len(out.shape) == 1:
-        out = numpy.stack((out[:-1], out[1:]), axis=1)
+        out = numpy.stack((out[:-1], out[1:]), axis=1).tolist()
 
     return out
 
```

Comment:
Shouldn't it just be the numpy array here rather than tolist()? out starts out as a numpy array.

Reply:
Oh, that was to normalize the types to plain Python. It's metadata, and although it's possible to iterate over NumPy arrays, it's less surprising for it to be Python `list` and `int`.

Also, the arrayification doesn't determine which integer type it is. If, for instance, it's an unsigned type, subtraction doesn't always work as expected. If it's int8, it could roll over. All sorts of bad things could happen that we just don't have to think about if it's turned into genetic `int`.

But being able to accept very generic types from users is also a benefit.

That's why I did that.

Reply:
Sure but now the code paths return different types.

Reply:
Yikes—it's supposed to be `.tolist()` for everything. That should have been on the `return`.

Are you in a position where you'd be able to do that easily? (I'm on a phone and my computer is shut down.)

Anyway, @ioanaif, this should be fixed before we merge.